// @griffithswaite/ts-plsql-parser/build/parser.js
import { CharStream, CommonTokenStream, } from "antlr4";
import PlSqlLexer from "./antlr/PlSqlLexer.js";
import PlSqlParser from "./antlr/PlSqlParser.js";
/**
 * Get a parser from a string containing PL/SQL code.
 * @param input String containing PL/SQL code
 * @returns PlSqlParser
 */
export function getParserFromInput(input) {
    const inputStream = new CharStream(input);
    let lexer = new PlSqlLexer(inputStream);
    let tokenStream = new CommonTokenStream(lexer);
    let parser = new PlSqlParser(tokenStream);
    return parser;
}
/**
 * Get a parser from a file containing PL/SQL code.
 * @param inputFile Path to file containing PL/SQL code
 * @returns PlSqlParser
 */
export async function getParserFromFile(inputFile) {
    const { FileStream } = await import("antlr4");
    const inputStream = new FileStream(inputFile);
    let lexer = new PlSqlLexer(inputStream);
    let tokenStream = new CommonTokenStream(lexer);
    let parser = new PlSqlParser(tokenStream);
    return parser;
}
/**
 * Given a multiline string, a character position, and a line number, return the line number
 * and character position of the character that is at the given position in the string.
 * @param str Entire input string
 * @param position Character position
 * @param line Line number
 * @returns [line number, character position]
 */
function getLineAndCharacterFromPosition(str, position, line) {
    let lineStart = 0;
    let lineEnd = str.indexOf("\n");
    let lineNum = 1;
    while (lineNum < line) {
        lineStart = lineEnd + 1;
        lineEnd = str.indexOf("\n", lineStart);
        lineNum++;
    }
    return [lineNum, position - lineStart + 1];
}
/**
 * Given a string containing PL/SQL code and a parser tree, return a nested tree of nodes
 * @param input
 * @param tree
 * @returns { nodes: ParsedNode[], stats: Record<string, number> }
 */
export function getParsedNodes(input, tree) {
    let nodeStats = {};
    const printNode = (node) => {
        let start = null;
        let stop = null;
        if (node.exception) {
            console.log(node.exception);
            return node.exception;
        }
        const type = node.constructor.name;
        nodeStats[type] = (nodeStats[type] || 0) + 1;
        if (node.start) {
            start = getLineAndCharacterFromPosition(input, node.start.start, node.start.line).join(":");
        }
        if (node.stop) {
            const stopPos = getLineAndCharacterFromPosition(input, node.stop.stop, node.stop.line);
            stop = `${stopPos[0]}:${stopPos[1] + 1}`;
        }
        return {
            type: type,
            text: node.start && node.stop
                ? input.substring(node.start.start, node.stop.stop + 1)
                : node.getText(),
            start: start,
            stop: stop,
            nodes: (node.children || []).map(printNode),
        };
    };
    return { nodes: [printNode(tree)], stats: nodeStats };
}
